File: lib/providers/url_resolver.dart
import 'package:flutter_riverpod/flutter_riverpod.dart' show Provider;
import 'package:kalinka/connection_settings_provider.dart'
    show connectionSettingsProvider;

class UrlResolver {
  UrlResolver(this.base);
  final String base;

  String abs(String path) {
    if (path.startsWith('http')) return path;
    final b = base.endsWith('/') ? base.substring(0, base.length - 1) : base;
    final p = path.startsWith('/') ? path : '/$path';
    return '$b$p';
  }

  String get baseKey => base; // stable key if you ever need it
}

// 2) Provider wired to your settings
final urlResolverProvider = Provider<UrlResolver>((ref) {
  // only rebuild resolver when the *base* actually changes
  final base = ref.watch(connectionSettingsProvider.select((s) => s.baseUrl));
  return UrlResolver(base.toString());
});

File: android/app/src/main/kotlin/com/example/kalinka/UrlResolver.kt
package com.example.kalinka

/**
 * URL resolver utility class for resolving relative paths against a base URL.
 */
class UrlResolver(private val base: String) {

    /**
     * Resolves a path to an absolute URL.
     * If the path is already absolute (starts with 'http'), returns it as-is.
     * Otherwise, combines it with the base URL.
     */
    fun abs(path: String): String {
        if (path.startsWith("http")) return path

        val b = if (base.endsWith("/")) {
            base.substring(0, base.length - 1)
        } else {
            base
        }

        val p = if (path.startsWith("/")) {
            path
        } else {
            "/$path"
        }

        return "$b$p"
    }

    /**
     * Returns the base URL as a stable key.
     */
    val baseKey: String
        get() = base
}
